## 2장 객체 생성과 파괴


------------------

#### 아이템 9 . try-finally 보다는 try-with-resources 를 사용하라

자바 라이브러리에는 close 메서드를 호출해 직접 닫아줘야 하는 자원이 많다. 자원 닫기는 클라이언트가 놓치기 쉬워서<br/>
예측할 수 없는 성능 문제로 이어지기도 한다. 이런 자원 중 상당수가 안전망으로 finalizer 를 활용하고는 있지만 <br/>
그리 믿을만 하지 못하다. 전통적으로 try-finally 가 예외와 return 경우에 닫힘을 보장하는 수단으로 쓰였다.

예외는 try 블록과 finally 블록 모두에서 발생할 수 있는데, try 가 중첩 사용되고 안쪽에서 에러가 난 뒤 close 조차 <br/>
에러가 난다면 두번째 예외가 첫번째 예외를 완전히 집어 삼켜버린다. 그럼 stack trace 에 첫번째 예외에 관한 정보는 <br/>
남지 않게 되어, 실제 시스템에서의 디버깅을 어렵게한다. 두번째 예외 대신 첫번째 예뢰를 기록하도록 코드를 수정할 수는 있지만<br/>
코드가 너무 지저분해져서 실제로 그렇게 까지 하는 경우는 거의 없다.

이러한 문제들은 자바 7 에 생긴 try-with-resources 덕에 모두 해결되었다. 이 구조를 사용하려면 해당 자원이 AutoCloseable<br/>
인터페이스를 구현해야한다. 이미 수많은 라이브러리들이 이것은 구현하거나 확장해뒀다.<br/>
try-with-resources 는 짧고 읽기 수월할 뿐 아니라 문제를 진단하기도 훨씬 좋다. close 에서 발생한 예외는 숨겨지고 <br/>
실제 발생한 예외가 기록된다. 숨겨진 예외들도 그냥 버려지지 않고 'suppressed' 꼬리표를 달고 출력된다. <br/>
또한, 자바 7에서 Throwable 에 추가 된 getSuppressed 메서드를 이용해 프로그램 코드에서 가져올 수도 있다.<br/>
게다가 try-with-resources 는 평범하게 중첩되지 않게 하나의 catch 절로 통합 할 수 있다.

> 핵심정리
> 꼭 회수해야하는 자원을 다룰때는 try-with-resources 를 사용하자. 예외는 없다.<br/>
> 코드는 더 짧고 분명해지고, 만들어지는 예외 정보도 훨씬 유용하다. 또한, 정확하고 쉽게 자원을 회수할 수 있다.